### PR TITLE
feat(logger): directive-based scope filtering and subscope support

### PR DESCRIPTION
**Pull request by goastler**
## Summary

Adds Rust-style per-scope log level filtering via \`PROSOPO_LOG_LEVEL\` and subscope support to \`with()\`.

### Directive syntax (extends existing env var)

\`PROSOPO_LOG_LEVEL\` now accepts both the existing bare level and a comma-separated directive string:

\`\`\`
PROSOPO_LOG_LEVEL=warn                          # unchanged — global floor
PROSOPO_LOG_LEVEL=warn,database=trace           # global warn, database:* gets trace
PROSOPO_LOG_LEVEL=provider:db=debug,http=info   # per-scope, no global default
\`\`\`

Scope segments use `:` as a separator. Matching is by most-specific prefix — \`provider:db\` matches a logger scoped \`provider:db:queries\` before falling back to \`provider\` and then the global default.

### Subscope in \`with()\`

\`\`\`ts
const logger = getLogger(LogLevel.enum.info, 'provider')
const dbLogger = logger.with({ requestId }, 'db')        // scope: "provider:db"
const queryLogger = dbLogger.with({ query }, 'queries')  // scope: "provider:db:queries"
\`\`\`

Child loggers inherit \`pretty\` and \`printStack\` settings from the parent. Their level is resolved from directives for the new scope, falling back to the parent's level.

### Runtime override

\`\`\`ts
import { setGlobalDirectives } from '@prosopo/logger'
setGlobalDirectives('warn,database=trace')  // useful in tests or dynamic config reload
\`\`\`

Directives are re-resolved at every \`print()\` call so \`setGlobalDirectives\` takes effect immediately without recreating loggers.

### Also fixes

- \`level\` field in emitted JSON records was recording the logger's configured threshold instead of the actual level of the message being printed

## Test plan

- [ ] \`PROSOPO_LOG_LEVEL=warn\` suppresses info/debug/trace globally
- [ ] \`PROSOPO_LOG_LEVEL=warn,database=trace\` gives trace on a logger scoped \`database\` while others stay at warn
- [ ] \`logger.with({}, 'sub')\` produces scope \`parent:sub\` and inherits pretty/printStack
- [ ] \`setGlobalDirectives\` takes effect on existing logger instances without recreation
- [ ] Bare \`PROSOPO_LOG_LEVEL=info\` still works as before (backward compat)

### CODE REVIEW

**Comment by goastler** [comment]
Fixed: replaced `process.env["PROSOPO_LOG_LEVEL"]` with `process.env.PROSOPO_LOG_LEVEL` to satisfy the `useLiteralKeys` biome rule. Added changeset.

File: packages/logger/src/logger.ts [hunk 1]
```diff
@@ -132,10 +211,15 @@ export class NativeLogger implements Logger {
 		return this.format;
 	}
 
-	with(obj: LogObject): Logger {
-		const newLogger = new NativeLogger(this.scope);
+	with(obj: LogObject, subscope?: string): Logger {
+		const newScope = subscope ? `${this.scope}:${subscope}` : this.scope;
+		const newLogger = new NativeLogger(newScope, this.levelMap);
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts:231]: Fixed: `with()` now checks `this.scope` before prepending — when the parent scope is empty the child scope is just `subscope` with no leading colon.

File: packages/logger/src/logger.ts [hunk 2]
```diff
@@ -132,10 +211,15 @@ export class NativeLogger implements Logger {
...
 		newLogger.defaultData = { ...this.defaultData, ...obj };
-		newLogger.setLogLevel(this.getLogLevel());
+		newLogger.setPretty(this.getPretty());
+		newLogger.setPrintStack(this.getPrintStack());
+		newLogger.setLogLevel(
+			resolveLevel(newScope, getGlobalDirectives(), this.getLogLevel()),
+		);
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts:237]: Acknowledged as a known design trade-off: the child logger's configured level is a snapshot at creation time. `print()` re-resolves from directives on every call, so runtime `setGlobalDirectives()` changes *do* take effect. The only case that doesn't update is if you call `setLogLevel()` directly on the child after construction, which is an explicit override. Documenting this in the JSDoc would be useful but is a follow-up.

File: packages/logger/src/logger.ts [hunk 3]
```diff
@@ -85,6 +86,84 @@ export function parseLogLevel(
...
+
+/**
+ * Parse a PROSOPO_LOG_LEVEL directive string into a scope→level map.
+ * An entry with an empty-string key represents the global default.
+ */
+export function parseDirectives(raw: string): Directives {
+	const map: Directives = new Map();
+	for (const part of raw.split(",")) {
+		const trimmed = part.trim();
+		if (!trimmed) continue;
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts:122]: Fixed: added unit tests for `parseDirectives`, `resolveLevel`, `setGlobalDirectives`, `parseLogLevel` (directive syntax), and `Logger.with` subscope (including empty-scope edge case) to `logger.unit.test.ts`.

File: packages/logger/src/logger.ts [hunk 4]
```diff
@@ -237,8 +321,12 @@ export class NativeLogger implements Logger {
...
 		level: LogLevel,
 	): void {
-		if (this.levelMap[level] > this.levelNum) {
-			return; // skip logging if the level is higher than the current log level
+		// Re-resolve at print time so directives set after construction are respected.
+		const effectiveLevelNum = this.levelMap[
+			resolveLevel(this.scope, getGlobalDirectives(), this.level)
+		];
+		if (this.levelMap[level] > effectiveLevelNum) {
+			return; // skip logging if the level is above the effective threshold
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts]: Fixed: removed the `levelNum` field from `NativeLogger`. `print()` uses `effectiveLevelNum` exclusively and `setLogLevel` no longer maintains the dead field.

File: packages/logger/src/logger.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts:237]: Same as the earlier sticky-level comment — print() re-resolves directives on every call so runtime changes take effect. The initial `setLogLevel` in `with()` gives the child its own starting level derived from current directives, which is the documented behaviour.

File: packages/logger/src/logger.ts [hunk 5]
```diff
@@ -85,6 +86,84 @@ export function parseLogLevel(
...
 }
 
+// ---------------------------------------------------------------------------
+// Directive-based filtering
+//
+// PROSOPO_LOG_LEVEL supports both a bare level and a comma-separated directive
+// string with optional per-scope overrides, e.g.:
+//   "warn"                         – global floor
+//   "warn,database=trace"          – global warn, database:* gets trace
+//   "database=trace,http=debug"    – per-scope only, no global default
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts:106]: Fixed: `parseLogLevel` now handles directive strings — it scans for the first segment without an `=` and parses that as the global level, so `parseLogLevel("warn,database=trace")` correctly returns `"warn"`.

File: packages/logger/src/logger.ts [hunk 6]
```diff
@@ -85,6 +86,84 @@ export function parseLogLevel(
...
+}
+
+let _globalDirectives: Directives | undefined;
+
+function getGlobalDirectives(): Directives {
+	if (!_globalDirectives) {
+		const raw =
+			typeof process !== "undefined"
+				? (process.env.PROSOPO_LOG_LEVEL ?? "")
+				: "";
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts:163]: Fixed: changed to `process.env?.PROSOPO_LOG_LEVEL` with optional chaining.

File: packages/logger/src/logger.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts:122]: Fixed: added unit tests for `parseDirectives`, `resolveLevel`, `setGlobalDirectives`, `parseLogLevel` (directive syntax), and `Logger.with` subscope.

File: packages/logger/src/logger.ts [hunk 7]
```diff
@@ -237,8 +321,13 @@ export class NativeLogger implements Logger {
...
 		level: LogLevel,
 	): void {
-		if (this.levelMap[level] > this.levelNum) {
-			return; // skip logging if the level is higher than the current log level
+		// Re-resolve at print time so directives set after construction are respected.
+		const effectiveLevelNum =
+			this.levelMap[
+				resolveLevel(this.scope, getGlobalDirectives(), this.level)
+			];
+		if (this.levelMap[level] > effectiveLevelNum) {
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts]: Fixed: removed the dead `levelNum` field — same as the earlier comment.

File: packages/logger/src/logger.ts [hunk 8]
```diff
@@ -34,11 +34,12 @@ export type Logger = {
...
 	log(level: LogLevel, fn: LogRecordFn): void;
 	/**
-	 * Creates a new logger instance which includes the given object in every log message. Akin to a child logger.
-	 * This is useful for adding context to log messages, such as user IDs, request IDs, etc.
-	 * @param obj An object to log, which will be added to every log message.
+	 * Creates a child logger with merged default data and an optional subscope appended to the
+	 * current scope (e.g. parent "database" + subscope "queries" → "database:queries").
+	 * The child's effective level is resolved from PROSOPO_LOG_LEVEL directives for the new scope,
+	 * falling back to the parent's level when no directive matches.
 	 */
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts:45]: Fixed: updated the with() JSDoc to describe the actual behaviour — the child's configured level is a snapshot resolved at creation, and print() re-resolves directives at emit time using that snapshot as the fallback (a direct setLogLevel() on the child overrides it).

File: packages/logger/src/tests/logger.unit.test.ts [hunk 9]
```diff
@@ -13,7 +13,14 @@
...
+import {
+	getLogger,
+	parseDirectives,
+	parseLogLevel,
+	resolveLevel,
+	setGlobalDirectives,
+	stringifyBigInts,
+} from "../logger.js";
 describe("stringifyBigInts", () => {
 	it("should strigify big int", () => {
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/tests/logger.unit.test.ts]: Fixed: typo corrected to 'stringify'.

File: packages/logger/src/tests/logger.unit.test.ts [hunk 10]
```diff
@@ -68,3 +75,95 @@ describe("stringifyBigInts", () => {
...
+describe("setGlobalDirectives", () => {
+	it("takes effect immediately on newly created loggers", () => {
+		setGlobalDirectives("warn,test-scope=trace");
+		const logger = getLogger("info", "test-scope");
+		// The logger starts at info but the directive says trace for test-scope.
+		// Effective level resolved in print(), so a trace log should still pass
+		// through (we just verify the scope is set).
+		expect(logger.getScope()).toBe("test-scope");
+		setGlobalDirectives(""); // reset
+	});
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/tests/logger.unit.test.ts]: Fixed: the directive-mutating tests now restore both console spies and global directives in finally/afterEach so state can't leak on failure.

File: packages/logger/src/logger.ts [hunk 11]
```diff
@@ -237,8 +330,13 @@ export class NativeLogger implements Logger {
...
 	): void {
-		if (this.levelMap[level] > this.levelNum) {
-			return; // skip logging if the level is higher than the current log level
+		// Re-resolve at print time so directives set after construction are respected.
+		const effectiveLevelNum =
+			this.levelMap[
+				resolveLevel(this.scope, getGlobalDirectives(), this.level)
+			];
+		if (this.levelMap[level] > effectiveLevelNum) {
+			return; // skip logging if the level is above the effective threshold
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts]: Fixed: added console.debug spy tests asserting per-scope directives actually suppress/emit, plus a console.error test asserting the JSON output's level field equals the message level rather than the configured threshold.

File: packages/logger/src/logger.ts [hunk 12]
```diff
@@ -135,10 +220,19 @@ export class NativeLogger implements Logger {
...
 	}
 
-	with(obj: LogObject): Logger {
-		const newLogger = new NativeLogger(this.scope);
+	with(obj: LogObject, subscope?: string): Logger {
+		const newScope = subscope
+			? this.scope
+				? `${this.scope}:${subscope}`
+				: subscope
+			: this.scope;
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/logger.ts]: Fixed: with() now trims the subscope (and treats whitespace-only as absent) so scopes match the trimmed keys from parseDirectives(). Added tests for both cases.

File: packages/logger/src/tests/logger.unit.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/tests/logger.unit.test.ts]: Fixed: the test now asserts observable behaviour — toggling directives at runtime suppresses then emits an existing logger's debug output via a console.debug spy — and restores console + directives in finally/afterEach.

File: packages/logger/src/tests/logger.unit.test.ts [hunk 13]
```diff
@@ -133,3 +139,147 @@ describe("stringifyBigInts", () => {
...
+		expect(resolveLevel("any:scope", d, "info")).toBe("info");
+	});
+});
+
+describe("setGlobalDirectives", () => {
+	// Always restore the global directives so a failing assertion can't leak
+	// state into later tests.
+	afterEach(() => {
+		setGlobalDirectives("");
+	});
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/tests/logger.unit.test.ts:207]: Fixed: the cleanup now restores `setGlobalDirectives(process.env.PROSOPO_LOG_LEVEL ?? "")` instead of forcing empty, so a non-empty PROSOPO_LOG_LEVEL in the surrounding environment isn't clobbered.

File: packages/logger/src/tests/logger.unit.test.ts [hunk 14]
```diff
@@ -133,3 +139,147 @@ describe("stringifyBigInts", () => {
...
+		} finally {
+			debugSpy.mockRestore();
+		}
+	});
+});
+
+describe("emitted record", () => {
+	afterEach(() => {
+		setGlobalDirectives("");
+	});
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/tests/logger.unit.test.ts:231]: Fixed: same restore-to-env-derived directives applied to this cleanup as well.

File: packages/logger/src/tests/logger.unit.test.ts [hunk 15]
```diff
@@ -133,3 +139,147 @@ describe("stringifyBigInts", () => {
...
+		const child = parent.with({}, "   ");
+		expect(child.getScope()).toBe("provider");
+	});
+
+	it("inherits default data from parent", () => {
+		const parent = getLogger("info", "test").with({ requestId: "abc" });
+		const child = parent.with({ extra: 1 });
+		// Can't inspect defaultData directly but verify scope is unchanged
+		expect(child.getScope()).toBe("test");
+	});
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/tests/logger.unit.test.ts]: Fixed: the 'inherits default data' test now spies on console.info and asserts the emitted record.data contains the parent's (`requestId`), child's (`extra`) and per-call (`perCall`) keys, so a broken merge would fail.

[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/tests/logger.unit.test.ts]: Fixed: rewrote the test to assert the merged default data in the emitted JSON record (requestId + extra + per-call data) via a console.info spy, rather than only checking scope.

File: packages/logger/src/tests/logger.unit.test.ts [hunk 16]
```diff
@@ -133,3 +139,147 @@ describe("stringifyBigInts", () => {
...
+});
```
[the remark this answers is not in the document]
Comment (goastler) [reply packages/logger/src/tests/logger.unit.test.ts]: Fixed: the test now logs via the child and asserts record.data matches { requestId: 'abc', extra: 1, perCall: true } from the console.info spy output.

File: .changeset/logger-subscope-trim.md [hunk 17]
```diff
@@ -0,0 +1,5 @@
+---
+"@prosopo/logger": patch
+---
+
+Trim whitespace from subscopes in Logger.with so they match directive keys, clarify with() JSDoc, and bump the @prosopo/fingerprintjs submodule to align its @prosopo/config dependency with the workspace (3.3.3).
```
[the remark this answers is not in the document]
Comment (goastler) [reply .changeset/logger-subscope-trim.md]: Fixed: split the changeset — `logger-subscope-trim.md` is now scoped to the logger changes only, and the fingerprintjs submodule/config alignment moved to a separate empty changeset (`fingerprintjs-config-align.md`, private package, no release).

File: .changeset/fingerprintjs-config-align.md [hunk 18]
```diff
@@ -0,0 +1,4 @@
+---
+---
+
```
[the remark this answers is not in the document]
Comment (goastler) [reply .changeset/fingerprintjs-config-align.md:3]: This is intentional and valid — an empty-frontmatter changeset is exactly what `@changesets/cli add --empty` produces, and changesets parses it as 'no release for any package'. It's the repo's sanctioned pattern for private-only changes (see the dependabot-changeset workflow). The `check changesets` CI job passes with it green, confirming it parses correctly. A private package can't be given a release entry anyway, so an empty changeset is the correct way to record the fingerprintjs submodule bump.
